### PR TITLE
Add wild hotplug/unplug case

### PR DESCRIPTION
**Pull request by qingwangrh**
ID:1943481
Signed-off-by: qingwangrh <qinwang@redhat.com>

### CODE REVIEW

**Comment by qingwangrh** [comment]
python ConfigTest.py --testcase=multi_disk_wild_hotplug --guestname=RHEL.8.4.0 --driveformat=virtio_scsi,virtio_blk --machine=q35,i440fx --clone=no  --iothread_scheme=roundrobin --nr_iothreads=2
(1/4) Host_RHEL.m8.u4.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.multi_disk_wild_hotplug: PASS (1173.15 s)
 (2/4) Host_RHEL.m8.u4.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.multi_disk_wild_hotplug.q35: PASS (1497.73 s)
 (3/4) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.multi_disk_wild_hotplug: PASS (2580.98 s)
 (4/4) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.multi_disk_wild_hotplug.q35: PASS (2585.03 s)

**Comment by qingwangrh** [comment]
@YongxueHong @zhenyzha Please help to review. thanks.

File: qemu/tests/multi_disk_wild_hotplug.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    :param env: Dictionary with test environment.
+    """
+
+    def _run_sg_luns():
+        script_cmd = r"""cat << EOF >/tmp/guest-run.sh
+trap 'kill \$(jobs -p)' EXIT SIGINT
+
+for i in \`seq 0 32\` ; do
+    while true ; do
+        sg_luns /dev/sdb > /dev/null 2>&1
```
Comment (YongxueHong) [inline qemu/tests/multi_disk_wild_hotplug.py:34]: `/dev/sdb` is correct when the` driver_format = virtio_blk`,
I think there is no `/dev/sd*` inside guest, just `/dev/vd*`.

File: qemu/tests/multi_disk_wild_hotplug.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    done &
+done
+echo 'wait'
+wait
+EOF
+chmod 755 /tmp/guest-run.sh
+cat /tmp/guest-run.sh
+"""
```
[the remark this answers is not in the document]
Comment (qingwangrh) [reply qemu/tests/multi_disk_wild_hotplug.py]: Updated

File: qemu/tests/multi_disk_wild_hotplug.py (hunk 1 above)
Comment (qingwangrh) [reply qemu/tests/multi_disk_wild_hotplug.py]: This script only run for virtio_scsi.

**Comment by zhenyzha** [comment]
# git am 2722.patch 
```
Applying: Add wild hotplug/unplug case
.git/rebase-apply/patch:57: new blank line at EOF.
+
warning: 1 line adds whitespace errors.

```

**Comment by qingwangrh** [comment]
> git am 2722.patch
> 
> ```
> Applying: Add wild hotplug/unplug case
> .git/rebase-apply/patch:57: new blank line at EOF.
> +
> warning: 1 line adds whitespace errors.
> ```

Fixed.

**Review by YongxueHong — APPROVED** [review]

File: qemu/tests/cfg/multi_disk_wild_hotplug.cfg [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    stg_image_num = 41
+    stg_image_size = 256M
+    repeat_num = 300
+    unplug_time = 3
+    plug_time = 3
+    virtio_scsi:
+      guest_operation = _run_sg_luns
+    virtio_blk:
+      repeat_num = 100
+      q35:
```
Comment (zhenyzha) [inline qemu/tests/cfg/multi_disk_wild_hotplug.cfg:21]: `22:46:39 WARNI| Ignore hotplug error: QMP command 'device_add' failed    (arguments: OrderedDict([('driver', 'virtio-blk-pci'), ('id', 'stg1'), ('drive', 'drive_stg1'), ('bootindex', '2'), ('write-cache', 'on'), ('bus', 'pcie-root-port-5'), ('addr', '0x0'), ('iothread', 'iothread0')]),    error message: {'class': 'GenericError', 'desc': "Duplicate ID 'stg1' for device"})`

need add ARM support
`q35, arm64-pci, arm64-mmio:`

File: qemu/tests/multi_disk_wild_hotplug.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            dev = vm.devices.get_by_qid(name)[0]
+            images_params[name] = copy.deepcopy(dev.params)
+
+    def _hotplug_images():
+        for i in range(1, image_num):
+            name = "stg%d" % i
+            try:
+                vm.monitor.cmd("device_add", images_params[name], debug=False)
+            except QMPCmdError as e:
+                logging.warning('Ignore hotplug error: %s', str(e))
```
Comment (zhenyzha) [inline qemu/tests/multi_disk_wild_hotplug.py:62]: After adding arm support, I hit the following error. It is suspected that the insertion is too fast.
Survey
`23:10:15 WARNI| Ignore hotplug error: QMP command 'device_add' failed    (arguments: OrderedDict([('driver', 'virtio-blk-pci'), ('id', 'stg1'), ('drive', 'drive_stg1'), ('bootindex', '2'), ('write-cache', 'on'), ('bus', 'pcie-root-port-5'), ('addr', '0x0'), ('iothread', 'iothread0')]),    error message: {'class': 'GenericError', 'desc': "Duplicate ID 'stg1' for device"})`

**Comment by zhenyzha** [comment]
Add PPC platform review:
@maxujun Please help to review. thanks.

File: qemu/tests/multi_disk_wild_hotplug.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+    def _hotunplug_images():
+        for i in range(1, image_num):
+            name = "stg%d" % i
+            try:
+                vm.monitor.cmd("device_del", {"id": name}, debug=False)
```
Comment (zhenyzha) [inline qemu/tests/multi_disk_wild_hotplug.py:68]: After adding'`time.sleep(1)`' to'`_hotplug_images()`' and'`_hotunplug_images()`' respectively, the above error was not hit again.

File: qemu/tests/multi_disk_wild_hotplug.py (hunk 4 above)
Comment (qingwangrh) [reply qemu/tests/multi_disk_wild_hotplug.py:62]: This error can be ignored for virtio_blk due to the timeout hard to determind. and virtio_blk is not the test point. 
I suggest you mentioned scenario just added under virtio_scsi

File: qemu/tests/multi_disk_wild_hotplug.py (hunk 5 above)
Comment (qingwangrh) [reply qemu/tests/multi_disk_wild_hotplug.py:68]: comment as above

**Comment by zhenyzha** [comment]
```
(1/1) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.aarch64.io-github-autotest-qemu.multi_disk_wild_hotplug.arm64-pci: PASS (2666.46 s)
(1/1) Host_RHEL.m8.u4.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.4.0.aarch64.io-github-autotest-qemu.multi_disk_wild_hotplug.arm64-pci: PASS (4659.65 s)
```

LGTM.Acked-by: Zhenyu Zhang <zhenyzha@redhat.com>

**Review by zhenyzha — APPROVED** [review]

**Comment by XueqiangWei** [comment]
@YongxueHong Please help review it again. Many thanks.

**Comment by YongxueHong** [comment]
> Add PPC platform review:
> @maxujun Please help to review. thanks.

Hi @maxujun 
Please help to review, thanks.
CC @XueqiangWei @qingwangrh

**Comment by maxujun** [comment]
(1/2) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.multi_disk_wild_hotplug: PASS (2710.43 s)
(2/2) Host_RHEL.m8.u4.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.multi_disk_wild_hotplug: ERROR: VM 'avocado-vt-vm1' failed to start: Error occurred while executing make_create_command(). Check the log for traceback. (12.66 s)
 Failed to insert device because of "ParentBus({'aobject': 'pci.0'}): No free matching bus" when using virtio_blk disk format.

**Comment by XueqiangWei** [comment]
@maxujun Please help review it again. Many thanks.

**Comment by maxujun** [comment]
(1/2) Host_RHEL.m8.u4.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.multi_disk_wild_hotplug: PASS (746.61 s)
 (2/2) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.multi_disk_wild_hotplug: PASS (2764.13 s)

LGTM，Ack

**Pull request merged**